src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1242,19 +1248,210 @@
+                              (catch Exception e
+                                ;; Using mapv here rather than map to avoid laziness, otherwise compilation failure might
+                                ;; fall into the throw below for the wrong reason.
+                                (mapv (fn [expr compilation-ctx]
+                                        (with-bindings
+                                          {#'*compile-ctx* (:compile-ctx compilation-ctx)
+                                           #'*file* (:file compilation-ctx *file*)}
+                                          (try-eval expr)))
+                                      exprs
+                                      compilation-ctxs)
+                                ;; If none of the rules are the issue, it is likely that the
+                                ;; size of the code trying to be evaluated has exceeded the limit
+                                ;; set by java.
+                                (throw (ex-info (str "There was a failure while batch evaling the node expressions, " \newline
+                                                     "but wasn't present when evaling them individually. This likely indicates " \newline
+                                                     "that the method size exceeded the maximum set by the jvm, see the cause for the actual error.")
+                                                {:compilation-ctxs compilation-ctxs}
+                                                e)))))]
+    (into {}
+          cat
+          ;; Grouping by ns, most expressions will not have a defined ns, only expressions from production nodes.
+          ;; These expressions must be evaluated in the ns where they were defined, as they will likely contain code
+          ;; that is namespace dependent.
+          (for [[nspace group] (sort-by key (group-by (comp :ns second val) key->expr))
+                ;; Partitioning the number of forms to be evaluated, Java has a limit to the size of methods if we were
+                ;; evaluate all expressions at once it would likely exceed this limit and throw an exception.
+                expr-batch (partition-all partition-size group)
+                :let [node-expr-keys (mapv first expr-batch)
+                      compilation-ctxs (mapv (comp second val) expr-batch)
+                      exprs (mapv (comp first val) expr-batch)]]
+            (mapv vector
+                  node-expr-keys
+                  (with-bindings (if nspace
+                                   {#'*ns* (the-ns nspace)}
+                                   {})
+                    (batching-try-eval compilation-ctxs exprs)))))))
+
+(defn safe-get
+  "A helper function for retrieving a given key from the provided map. If the key doesn't exist within the map this
+   function will throw an exception."
+  [m k]
+  (let [v (get m k ::not-found)]
+    (if (= v ::not-found)
```

Comment:
To @WilliamParker 's point:

> The potentially unreliable use of identical? in the safe-get function.

The likelihood of this changing in clj is very slim. It is even slimmer that it'd change for keyword literals. However, if choosing to remain paranoid for a change like that to happen, I still think an `identical?`-based sentinel test is what would typically be done in this sort of case

```clj
(let [not-found ::not-found
       v (get m k not-found)]
  (if (identical? v not-found) <etc>))
```

Alternatively, I've even seen the more aggressive, guaranteed-unique sentinel used:
```clj
(let [not-found (Object.)      ;; <---- unique object identity
       v (get m k not-found)]
  (if (identical? v not-found) <etc>))
```
Using a keyword feels more idiomatic, but using the Object is a stronger guarantee, but feels more like idiomatic Java. 

I am not against the keyword style, but in a generic fn like `safe-get`, I'd lean towards using an approach where `identical?` is used instead of `=`.

Reply:
Yeah, I agree that the likelihood of it changing is low, but not entirely trivial.  I'd rather not take the risk when there is an easy way not to though.  I'm OK with using the same reference to enforce identity as you describe.